File: output_20260320_065752/gpt52/c/Struct/account.c
#include "stdlib.h"

struct account
{
    int balance;
};

/*@ predicate account(struct account *a; int bal) =
        a != 0 &*& malloc_block_account(a) &*& a->balance |-> bal;
@*/

struct account *create_account()
//@ requires true;
//@ ensures account(result, 0);
{
    struct account *myAccount = malloc(sizeof(struct account));
    if (myAccount == 0)
    {
        abort();
    }
    //@ close account(myAccount, _);
    myAccount->balance = 0;
    //@ open account(myAccount, _);
    //@ close account(myAccount, 0);
    return myAccount;
}

void account_set_balance(struct account *myAccount, int newBalance)
//@ requires account(myAccount, _);
//@ ensures account(myAccount, newBalance);
{
    //@ open account(myAccount, _);
    myAccount->balance = newBalance;
    //@ close account(myAccount, newBalance);
}

void account_dispose(struct account *myAccount)
//@ requires account(myAccount, _);
//@ ensures true;
{
    //@ open account(myAccount, _);
    free(myAccount);
}

int main()
//@ requires true;
//@ ensures true;
{
    struct account *myAccount = create_account();
    account_set_balance(myAccount, 5);
    account_dispose(myAccount);
    return 0;
}

File: output/gpt52/rust/purely_unsafe/solutions/account.rs
use std::alloc::{alloc, dealloc, handle_alloc_error, Layout};

struct Account {
    balance: i32,
}

/*@

predicate account(p: *mut Account; bal: i32) =
    std::alloc::alloc_block(p as *mut u8, Layout::new::<Account>()) &*&
    struct_Account(p, bal);

@*/

impl Account {
    #[requires(true)]
    #[ensures(result != std::ptr::null_mut::<Account>() &*& account(result, 0))]
    unsafe fn create() -> *mut Account {
        let my_account = alloc(Layout::new::<Account>()) as *mut Account;
        if my_account.is_null() {
            handle_alloc_error(Layout::new::<Account>());
        }
        //@ close struct_Account(my_account, _);
        (*my_account).balance = 0;
        //@ close account(my_account, 0);
        my_account
    }

    #[requires(account(my_account, _))]
    #[ensures(account(my_account, new_balance))]
    unsafe fn set_balance(my_account: *mut Account, new_balance: i32) {
        //@ open account(my_account, _);
        //@ open struct_Account(my_account, _);
        (*my_account).balance = new_balance;
        //@ close struct_Account(my_account, new_balance);
        //@ close account(my_account, new_balance);
    }

    #[requires(account(my_account, _))]
    #[ensures(true)]
    unsafe fn dispose(my_account: *mut Account) {
        //@ open account(my_account, _);
        //@ open struct_Account(my_account, _);
        //@ open std::alloc::alloc_block(my_account as *mut u8, Layout::new::<Account>());
        dealloc(my_account as *mut u8, Layout::new::<Account>());
    }
}

#[requires(true)]
#[ensures(true)]
fn main() {
    unsafe {
        let my_account = Account::create();
        Account::set_balance(my_account, 5);
        Account::dispose(my_account);
    }
}